server/sonar-server/src/main/java/org/sonar/server/authentication/JwtTokenUpdater.java

```diff
@@ -0,0 +1,144 @@
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.lang.time.DateUtils;
+import org.sonar.api.platform.Server;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.System2;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.user.UserDto;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Arrays.asList;
+
+@ServerSide
+public class JwtTokenUpdater {
+
+  private static final Logger LOG = Loggers.get(GenerateJwtTokenFilter.class);
+
+  private static final String JWT_COOKIE = "JWT-SESSION";
+
+  // This timeout is used to disconnect the user we he has not browse any page for a while
+  private static final int SESSION_TIMEOUT_IN_SECONDS = 20 * 24 * 60 * 60;
+
+  // This refresh time is used to generate a new session
+  private static final int SESSION_REFRESH_IN_MINUTES = 5;
+
+  private static final String RAILS_USER_ID_SESSION = "user_id";
+
+  private final System2 system2;
+  private final DbClient dbClient;
+  private final Server server;
+  private final JwtToken jwtToken;
+
+  public JwtTokenUpdater(System2 system2, DbClient dbClient, Server server, JwtToken jwtToken) {
+    this.jwtToken = jwtToken;
+    this.server = server;
+    this.dbClient = dbClient;
+    this.system2 = system2;
+  }
+
+  void createNewJwtToken(String userLogin, HttpServletResponse response) {
+    String token = jwtToken.encode(JwtToken.Jwt.builder()
+      .setUserLogin(userLogin)
+      .setExpirationTimeInSeconds(SESSION_TIMEOUT_IN_SECONDS)
+      .build());
+
+    LOG.trace("Create session for {}", userLogin);
+    response.addCookie(createCookie(JWT_COOKIE, token, SESSION_TIMEOUT_IN_SECONDS));
+  }
+
+  void validateJwtToken(HttpServletRequest request, HttpServletResponse response) {
+    Optional<Cookie> jwtCookie = findCookie(JWT_COOKIE, request);
+    if (jwtCookie.isPresent()) {
+      validateJwtToken(jwtCookie.get(), request, response);
+    }
+  }
+
+  private void validateJwtToken(Cookie jwtCookie, HttpServletRequest request, HttpServletResponse response) {
+    String value = jwtCookie.getValue();
+    checkNotNull(value, "JWT cookie is null");
+    Optional<Claims> claims = jwtToken.decode(value);
+    Optional<UserDto> user = claims.isPresent() ? getUser(claims.get().getSubject()) : Optional.empty();
+    if (claims.isPresent() && user.isPresent()) {
+      refreshSession(claims.get(), user.get(), request, response);
+    } else {
+      removeSession(request, response);
+    }
+  }
+
+  private void refreshSession(Claims token, UserDto user, HttpServletRequest request, HttpServletResponse response) {
+    String userLogin = token.getSubject();
+    LOG.trace("Validate session of {}", userLogin);
+    request.getSession().setAttribute(RAILS_USER_ID_SESSION, user.getId());
```

Comment:
In GenerateJwtTokenFilter, as we are after rails, we know that the user_id is set.
If we set it here again, it will required a DB call (I agree that it's not a big deal as it's only during /sessions/login) but you're right we can add it here too.